packages/transformers/js/core/src/side_effects.rs

```diff
@@ -0,0 +1,714 @@
+          self.has_side_effects = true;
+          return;
+        }
+
+        assign_expr.right.visit_with(self);
+      }
+
+      Expr::Call(call_expr) => {
+        let mut is_pure_function = false;
+
+        self.comments.with_leading(call_expr.span.lo, |comments| {
+          is_pure_function = comments
+            .into_iter()
+            .any(|comment| comment.text == "#__PURE__");
+        });
+
+        if !is_pure_function {
+          self.has_side_effects = true;
+        }
+      }
+      Expr::Arrow(arrow) => {
+        arrow.params.visit_with(self);
+      }
+      _ => {
+        self.has_side_effects = true;
+      }
+    }
+  }
+
+  fn visit_prop_name(&mut self, node: &PropName) {
+    match &node {
+      PropName::Computed(computed) => {
+        computed.expr.visit_with(self);
+      }
+      _ => {
+        // All other key nodes are safe
+      }
+    }
+  }
+}
+
+#[test]
+fn empty_filke() {
+  let code = r#""#;
+
+  assert_eq!(check_side_effects(code, vec![]), true);
+}
+
+#[test]
+fn barrel_file() {
+  let code = r#"
+    export * from './something';
+    export { a, b as c } from './something-else';
+  "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn function() {
+  let code = r#"
+      function a() {
+        console.log();
+      }
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn class() {
+  let code = r#"
+     class A {
+        constructor() {}
+     }
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn export_function() {
+  let code = r#"
+     export function a() {
+        console.log();
+     }
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn export_default_function() {
+  let code = r#"
+     export default function a() {
+        console.log();
+     }
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn export_default_local() {
+  let code = r#"
+    const thing = '';
+    export default thing;
+  "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn export_class() {
+  let code = r#"
+     export class A {
+        constructor() {}
+     }
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn export_default_class() {
+  let code = r#"
+     export default class A {
+        constructor() {}
+     }
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn export_named() {
+  let code = r#"
+     export {
+        a,
+        b as c
+     }
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn export_const_decalation() {
+  let code = r#"
+     export const a = '';
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), false);
+}
+
+#[test]
+fn side_effect_import() {
+  let code = r#"
+     import './some-file';
+    "#;
+
+  assert_eq!(check_side_effects(code, vec![]), true);
```

Comment:
I wasn't sure about this scenario. Should this be considered a side effect or not?

Reply:
Unless I'm missing something, I think you can't tell unless you know what happens in `./some-file`?  Or does it not matter because `import foo from './some-file'` would cause `./some-file` to execute just the same?

Reply:
What does Webpack do (and does it traverse imports in the analysis)?
I wouldn't expect there to be a difference regarding side-effects between `import "some-file"` and `import "foo" from "some-file"; export function bar(){ return foo; }` because they both *might* have side-effects in the imports (but again, maybe Webpack has run into this before)

(At least so far with explicit side-effects declarations, the meaning was "If you don't use any values from an imported side-effect-free asset, then the asset and its whole subgraph can be excluded". So this is a very valid concern)

Reply:
[Webpack doesn't mark this as a side effect](https://github.com/webpack/webpack/blob/853bfda35a0080605c09e1bdeb0103bcb9367a10/lib/optimize/SideEffectsFlagPlugin.js#L200C26-L200C26). I agree that there shouldn't be a difference between the two. My only concern was writing an import in this way is like the user declaring there intent that they want this import loaded now.